File: reprise.rb
#!/usr/bin/env ruby
%w(rubygems
   bluecloth
   rubypants
   haml
   sass
   stringio
   mailread
   time).each { |lib| require lib }

TITLE = 'Redflavor Journal'
AUTHOR = { :name => 'Eivind Uggedal',
           :email => 'eu@redflavor.com',
           :url => 'http://redflavor.com' }
ANALYTICS = 'UA-1857692-3'
PUBLIC = File.join(File.dirname(__FILE__), 'public')
ASSETS = File.join(File.dirname(__FILE__), 'assets')

# Format of time objects.
class Time
  def to_s
    self.strftime('%Y-%m-%d')
  end
end

# Stolen from Sinatra
def templates
  templates = {}

  eof = IO.read(caller.first.split(':').first).split('__FILE__').last
  data = StringIO.new(eof)

  current_template = nil
  data.each do |line|
    if line =~ /^##\s?(.*)/
      current_template = $1.to_sym
      templates[current_template] = ''
    elsif current_template
      templates[current_template] << line
    end
  end
  templates
end

def slugify(string)
  string.gsub(/[^\w\s-]/, '').gsub(/\s+/, '-').downcase
end

def htmlify(text)
  RubyPants.new(BlueCloth.new(text).to_html).to_html
end

def entries
  files = Dir[File.dirname(__FILE__) + '/entries/*'].sort.reverse
  files.collect do |file|
    meta_from_file(file)
  end
end

def meta_from_file(file)
  msg = Mail.new(file)
  tags = msg['Tags'].split
  filename = File.basename(file)
  results = filename.scan(/([\d]{4}).(\d\d).(\d\d)\.(.+)/).first
  date = Time.local(*results[0..2])
  title = results[3].gsub(/\./, ' ')

  { :body => msg.body.join(""),
    :tags => tags,
    :filename => filename,
    :date => date, 
    :title => title,
    :slug => slugify(title) }
end

def write_file(fname, data, root=PUBLIC)
  File.open(File.join(root, fname), 'w') { |f| f.puts data }
end

def create_dir(dirname, root=PUBLIC)
  FileUtils.mkdir_p File.join(root, dirname)
end

def clean_public
  FileUtils.rm_r PUBLIC if File.exists? PUBLIC
  FileUtils.mkdir_p PUBLIC
end

def generate_style
  style = Sass::Engine.new(templates[:style]).render
  write_file('style.css', style)
end

def render_haml(template, bind=binding)
  Haml::Engine.new(templates[:layout], {:format => :html4}).render(bind) do
    Haml::Engine.new(templates[template], {:format => :html4}).render(bind)
  end
end

def generate_fourofour
  fourofour = render_haml(:fourofour, binding)
  write_file('404.html', fourofour)
end

def generate_index
  @entries = entries
  index = render_haml(:index, binding)
  write_file('index.html', index)
end

def generate_tag_indexes
  all_entries = entries
  all_tags = all_entries.collect { |entry| entry[:tags] }.flatten.uniq
  all_tags.each do |tag_slug|
    @entries = all_entries.select { |entry| entry[:tags].include? tag_slug }
    @active_tag = tag_slug
    tag_index = render_haml(:index, binding)
    create_dir("/tags/#{tag_slug}")
    write_file("/tags/#{tag_slug}/index.html", tag_index)
  end
end

def generate_entries
  entries.each do |entry|
    @entry = entry
    @title = "#{TITLE}: #{@entry[:title]}"
    rendered = render_haml(:entry, binding)
    write_file("#{@entry[:slug]}.html", rendered)
  end
end


def distribute_assets
  FileUtils.cp_r "#{ASSETS}/.", PUBLIC
end

if __FILE__ == $0
  clean_public
  generate_style
  generate_fourofour
  generate_index
  generate_tag_indexes
  generate_entries
  distribute_assets
end

__END__

## layout
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
%html
  %head
    %title= @title ? @title : TITLE
    %meta{ 'http-equiv' => 'Content-Type', :content => 'text/html;charset=utf-8' }
    %link{ :rel => 'stylesheet', :type => 'text/css', :href => '/style.css' }
    %link{ :rel => 'alternate', :type => 'application/atom+xml', :title => TITLE, :href => 'http://feeds.feedburner.com/redflavor' }
  %body
    = yield
  %script{ :type => 'text/javascript'}
    var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
    document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
  %script{ :type => 'text/javascript'}
    = "var pageTracker = _gat._getTracker(\"#{ANALYTICS}\");"
    pageTracker._initData();
    pageTracker._trackPageview();

## index
%h1= TITLE
%address.author.vcard
  %a.url.fn{ :href => AUTHOR[:url] }= AUTHOR[:name]
  %br
  %a.email{ :href => "mailto:#{AUTHOR[:email]}" }= AUTHOR[:email]
- @entries.each_with_index do |entry, i|
  .hentry
    %abbr.updated{ :title => entry[:date].iso8601 }= entry[:date]
    %h2
      %a.entry-title{ :href => "/#{entry[:slug]}", :rel => 'bookmark' }
        = entry[:title]
    - if i == 0
      %ul.tags
        - entry[:tags].each do |tag|
          %li
          - if tag == @active_tag
            %a.active{ :href => "/tags/#{tag}", :rel => 'tag' }= tag
          - else
            %a{ :href => "/tags/#{tag}", :rel => 'tag' }= tag
      .entry-content~ htmlify(entry[:body])
    - else
      %ul.tags.inline
        - entry[:tags].each do |tag|
          %li
            - if tag == @active_tag
              %a.active{ :href => "/tags/#{tag}", :rel => 'tag' }= tag
            - else
              %a{ :href => "/tags/#{tag}", :rel => 'tag' }= tag

## entry
%h1
  %a{ :href => '/' }= TITLE
%address.author.vcard
  %a.url.fn{ :href => AUTHOR[:url] }= AUTHOR[:name]
  %br
  %a.email{ :href => "mailto:#{AUTHOR[:email]}" }= AUTHOR[:email]
.hentry
  %abbr.updated{ :title => @entry[:date].iso8601 }= @entry[:date]
  %h2
    %span.entry-title= @entry[:title]
  %ul.tags
    - @entry[:tags].each do |tag|
      %li
        %a{ :href => "/tags/#{tag}", :rel => 'tag' }= tag
  .entry-content~ htmlify(@entry[:body])

## fourofour
%h1
  %a{ :href => '/' }= TITLE
%address.author.vcard
  %a.url.fn{ :href => AUTHOR[:url] }= AUTHOR[:name]
  %br
  %a.email{ :href => "mailto:#{AUTHOR[:email]}" }= AUTHOR[:email]
%p
  Resource not found. Go back to
  %a{ :href => '/' } the front
  page.

## style
body
  :font-size 90%
  :font-family 'DejaVu Sans', 'Bitstream Vera Sans', Verdana, sans-serif
  :line-height 1.5
  :padding 0 10em 0 10em
  :width 40em
abbr.updated, ul.tags
  :float left
abbr.updated
  :border 0
  :margin 0.3em 0 0 -7em
ul.tags
  :list-style-type none
  :margin 3em 0 0 -7em
ul.tags.inline
  :float none
  :margin 0
ul.tags.inline li
  :display inline
ul.tags a.active
  :background #fcc
ul, ol
  :padding 0
blockquote
  :font-style italic
  :margin 0
blockquote em
  :font-weight bold
a
  :background #ffc
  :color #000
h1, address
  :font-style normal
  :text-align center
address
  :margin 0 0 2em 0
img
  :margin 1em 0 1em 0
h1, h2, h3, abbr.updated, address
  :font-family Georgia, 'DejaVu Serif', 'Bitstream Vera Serif', serif
  :font-style normal
  :font-weight normal
  a
    :background transparent
p
  :margin-bottom 0
p + p
  :margin-top 0
  :text-indent 1.1em
pre > code
  :border 0.15em solid #eee
  :border-left 1em solid #eee
  :display block
  :font-family 'DejaVu Sans Mono', 'Bitstream Vera Sans Mono', "Lucida Console", "monospaced
  :padding 1em 1em 1em 2em

File: reprise.py
#!/usr/bin/env python

from __future__ import with_statement

import os
import re
import time
import email
import shutil

import markdown

from os.path import abspath, realpath, dirname, join
from datetime import datetime, timedelta
from textwrap import dedent
from pygments.formatters import HtmlFormatter
from smartypants import smartyPants
from jinja2 import DictLoader, Environment
from lxml.builder import ElementMaker
from lxml.etree import tostring

TITLE = 'bytexbyte'
URL = 'http://bytexbyte.com'
STYLESHEET = 'styles.css'

AUTHOR = {
    'name': 'Matt Dawson',
    'url': 'http://bytexbyte.com',
    'elsewhere': {
        'Dawsoning': 'http://dawsoning.com/',
        'delicious': 'http://delicious.com/matthewtdawson/',
        'Facebook': 'http://facebook.com/mattdawson/',
        'flickr': 'http://flickr.com/photos/matthewtdawson/',
        'The Nested Float': 'http://thenestedfloat.com/',
        'twitter': 'http://twitter.com/mattdawson/',
    }
}

ROOT = abspath(dirname(__file__))
DIRS = {
    'source': join(ROOT, 'entries'),
    'build': join(ROOT, 'build'),
    'public': join(ROOT, 'public'),
    'assets': join(ROOT, 'assets'),
}

CONTEXT = {
    'author': AUTHOR,
    'stylesheet': STYLESHEET,
    'head_title': "%s" % TITLE,
}

def _markdown(content):
    return markdown.markdown(content, ['codehilite', 'def_list'])

def read_and_parse_entries():
    files = sorted([join(DIRS['source'], f)
                    for f in os.listdir(DIRS['source'])], reverse=True)
    entries = []
    for file in files:
        match = META_REGEX.findall(file)
        if len(match):
            meta = match[0]
            with open(file, 'r') as open_file:
                msg = email.message_from_file(open_file)
                date = datetime(*[int(d) for d in meta[0:3]])
                entries.append({
                    'slug': slugify(meta[3]),
                    'title': meta[3].replace('.', ' '),
                    'tags': msg['Tags'].split(),
                    'date': {'iso8601': date.isoformat(),
                             'rfc3339': rfc3339(date),
                             'display': date.strftime('%B %d, %Y'),},
                    'content_html': smartyPants(_markdown(msg.get_payload())),
                })
    return entries

def generate_index(entries, template):
    feed_url = "%s/index.atom" % URL
    html = template.render(dict(CONTEXT, **{'entries': entries,
                                            'feed_url': feed_url}))
    write_file(join(DIRS['build'], 'index.html'), html)
    atom = generate_atom(entries, feed_url)
    write_file(join(DIRS['build'], 'index.atom'), atom)

def generate_tag_indices(entries, template):
    for tag in set(sum([e['tags'] for e in entries], [])):
        tag_entries = [e for e in entries if tag in e['tags']]
        feed_url = "%s/tags/%s.atom" % (URL, tag)
        html = template.render(
            dict(CONTEXT, **{'entries': tag_entries,
                             'active_tag': tag,
                             'feed_url': feed_url,
                             'head_title': "%s: %s" % (CONTEXT['head_title'],
                                                       tag),}))
        write_file(join(DIRS['build'], 'tags', '%s.html' % tag), html)
        atom = generate_atom(tag_entries, feed_url)
        write_file(join(DIRS['build'], 'tags', '%s.atom' % tag), atom)

def generate_details(entries, template):
    for entry in entries:
        html = template.render(
            dict(CONTEXT, **{'entry': entry,
                             'head_title': "%s: %s" % (CONTEXT['head_title'],
                                                       entry['title'])}))
        write_file(join(DIRS['build'], '%s.html' % entry['slug']), html)

def generate_404(template):
        html = template.render(CONTEXT)
        write_file(join(DIRS['build'], '404.html'), html)

def generate_style(css):
    css2 = HtmlFormatter(style='trac').get_style_defs()
    write_file(join(DIRS['build'], STYLESHEET), ''.join([css, "\n\n", css2]))

def generate_atom(entries, feed_url):
    A = ElementMaker(namespace='http://www.w3.org/2005/Atom',
                     nsmap={None : "http://www.w3.org/2005/Atom"})
    entry_elements = []
    for entry in entries:
        entry_elements.append(A.entry(
            A.id(atom_id(entry=entry)),
            A.title(entry['title']),
            A.link(href="%s/%s" % (URL, entry['slug'])),
            A.updated(entry['date']['rfc3339']),
            A.content(entry['content_html'], type='html'),))
    return tostring(A.feed(A.author( A.name(AUTHOR['name']) ),
                           A.id(atom_id()),
                           A.title(TITLE),
                           A.link(href=URL),
                           A.link(href=feed_url, rel='self'),
                           A.updated(entries[0]['date']['rfc3339']),
                           *entry_elements), pretty_print=True)

def write_file(file_name, contents):
    with open(file_name, 'w') as open_file:
        open_file.write(contents.encode("utf-8"))

def slugify(str):
    return re.sub(r'\s+', '-', re.sub(r'[^\w\s-]', '',
                                      str.replace('.', ' ').lower()))

def atom_id(entry=None):
    domain = re.sub(r'http://([^/]+).*', r'\1', URL)
    if entry:
        return "tag:%s,%s:/%s" % (domain, entry['date']['display'],
                                  entry['slug'])
    else:
        return "tag:%s,2009-03-04:/" % domain

def rfc3339(date):
    offset = -time.altzone if time.daylight else -time.timezone
    return (date + timedelta(seconds=offset)).strftime('%Y-%m-%dT%H:%M:%SZ')

def get_templates():
    templates = {
    'base.html': """
    <!DOCTYPE html>
    <html lang="en">
      <head>
        <title>{{ head_title }}</title>
        <meta charset="UTF-8">
        <link rel="stylesheet" type="text/css" href="/{{ stylesheet }}">
        <link rel="alternate" type="application/atom+xml"
              title="{{ head_title }}" href="{{ feed_url }}">
        <link rel="openid.server" href="http://www.myopenid.com/server" />
        <link rel="openid.delegate" href="http://mattdawson.myopenid.com/" />
      </head>
      <body>
        <a href="http://github.com/mattd"><img class="ribbon"
        src="http://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png"
        alt="Fork me on GitHub"></a>
        <section id="site-info">
          <header>
            <h2>busy building</h2>
            <h1><a href="/">byte<span>x</span>byte</a></h1>
          </header>
          <article>
            <p>is the programming journal of <a
            href="http://www.google.com/profiles/matthewtdawson">
            {{ author.name }}</a>, a professional web developer and tech junkie 
            from Charlottesville, VA USA.</p>
            <p>Wanna get in touch? Email matt at this domain dot com.  I'll be 
            pleasant. Promise.</p>
            <p><strong>Elsewhere:</strong>
            {% for service, url in author.elsewhere.items() %}
              <a href="{{ url }}">{{ service }}</a>{% if not loop.last %},
              {% endif %}
            {% endfor %}
            </p>
          </article>
        </section><!--/#site-info-->
        {% block content %}
        {% endblock %}
        <footer>
          <nav>
            <strong>feeds:</strong> <a href="{{ feed_url }}">atom</a>
          </nav>
          <p>&copy; {{ author.name }}. 
          <a href="http://github.com/uggdeal/reprise/"> Reprise</a> powered.
          Hosted in <a href="http://rackspacecloud.com"> the cloud</a>.</p>
        </footer>
      </body>
    </html>
    """,

    'list.html': """
    {% extends "base.html" %}
    {% block content %}
      <section id="previews">
      {% for entry in entries %}
        {% include '_preview.html' %}
      {% endfor %}
      </section><!--/#previews-->
    {% endblock %}
    """,

    'detail.html': """
    {% extends "base.html" %}
    {% block content %}
      <section id="writing">
        {% include '_entry.html' %}
      </section><!--/#writing-->
    {% endblock %}
    """,

    '_preview.html': """
    <article class="clearfloat">
      <header>
        <time datetime="{{ entry.date.iso8601 }}">
          {{ entry.date.display }}
        </time>
        <h2><a href="/{{ entry.slug }}" rel="bookmark">
          {{ entry.title }}
        </a></h2>
      </header>
      <section class="body">
        <p>{{ entry.content_html|striptags|truncate(300) }} //
        <a href="/{{ entry.slug }}">Continue</a>.</p>
        <p class="tags"><strong>Tags:</strong>
          {% for tag in entry.tags %}
            <a href="/tags/{{ tag }}" rel="tag">{{ tag }}</a>{% if not loop.last %},
            {% endif %}
          {% endfor %}
        </p>
      </section><!--/.body-->
    </article>
    """,

    '_entry.html': """
    <article class="clearfloat">
      <header>
        <time datetime="{{ entry.date.iso8601 }}">
          {{ entry.date.display }}
        </time>
        <h2><a href="/{{ entry.slug }}" rel="bookmark">
          {{ entry.title }}
        </a></h2>
      </header>
      <section class="body">
        {{ entry.content_html }}
        <p class="tags"><strong>Tags:</strong>
          {% for tag in entry.tags %}
            <a href="/tags/{{ tag }}" rel="tag">{{ tag }}</a>{% if not loop.last %},
            {% endif %}
          {% endfor %}
        </p>
      </section><!--/.body-->
    </article>
    """,

    '404.html': """
    {% extends "base.html" %}
    {% block content %}
      <p>Resource not found. Go back to <a href="/">the front</a> page.</p>
    {% endblock %}
    """,

    STYLESHEET: """
    /* ----- the reset ----- */

    html, body, div, span, applet, object, iframe,
    h1, h2, h3, h4, h5, h6, p, blockquote, pre,
    a, abbr, acronym, address, big, cite, code,
    del, dfn, em, font, img, ins, kbd, q, s, samp,
    small, strike, strong, sub, sup, tt, var,
    b, u, i, center,
    dl, dt, dd, ol, ul, li,
    fieldset, form, input, label, legend,
    table, caption, tbody, tfoot, thead, tr, th, td {
        margin: 0;
        padding: 0;
    }

    /* ----- the basics ----- */

    body {
        background:#0c0c0c;
        color:#fff;
        font-family: 'Helvetica Neue', helvetica, arial, sans-serif;
        font-size:16px;
        padding-top:20px;
    }

    p {
        line-height:28px;
        margin-top:28px;
        margin-bottom:28px;
    }

    a {
        color:#fff;
        outline:none;
    }

    a:hover {
        background:#bc0d3c;
        text-decoration:none;
    }

    pre {
        background:#000;
        line-height:28px;
    }

    pre code {
        display:block;
        padding:24px 20px 32px;
        overflow:scroll;
    }

    code {
        background:#000;
        padding:0 5px;
    }

    abbr {
        border-bottom:1px dotted #fff;
        cursor:help;
    }

    blockquote {
        background:#161616;
        font-family:georgia,serif;
        font-style:italic;
        padding:28px 34px;
    }

    blockquote p {
        margin:0;	
    }

    /* ----- the sidebar ----- */

    #site-info {
        position:absolute;
        top:60px;
        left:720px;
        overflow:hidden;
        width:360px;
    }

    #site-info header h1 {
        font-size:76px;
        line-height:84px;
        margin:33px 0 -7px;
    }

    #site-info header h1 a {
        text-decoration:none;
        text-shadow:0 2px 2px #000;
    }

    #site-info header h1 a:hover {
        background:none;
        color:#cd0053;
    }

    #site-info header h1 a span {
        color:#cd0053;
    }

    #site-info header h1 a:hover span {
        color:#fff;
    }

    #site-info header h2 {
        color:#8c8c8c;
        font-size:24px;
        line-height:28px;
        margin:2px 0 -40px;
    }

    #site-info article {
        color:#8c8c8c;
        line-height:1.8em;
    }

    /* ----- the articles ----- */

    #writing,
    #previews {
        background:#1c1c1c;
        border-right:20px solid #161616;
        border-left:20px solid #161616;
        display:block;
        height:100%;
        margin:0 0 0 20px;
        padding:28px 36px 56px 0;
        width:604px;
    }

    #writing article,
    #previews article {
        clear:both;
        display:block;
    }

    #writing article header h2 a,
    #previews article header h2 a {
        background:#b7003d;	
        border-left:20px solid #920031;
        color:#fff;
        float:left;
        font-size:42px;
        line-height:56px;
        margin-bottom:20px;
        margin-left:-20px;
        padding:2px 20px 6px;
        text-decoration:none;
        text-shadow:0 2px 2px #0c0c0c;
    }

    #writing article header h2 a:hover,
    #previews article header h2 a:hover {
        background:#920031;
        border-left:20px solid #6e0025;
    }

    #writing article header time,
    #previews article header time {
        color:#8c8c8c;
        display:block;
        line-height:28px;
        margin:16px 0 12px 20px;
        text-shadow:0 1px 1px #0c0c0c;
    }

    #writing article section,
    #previews article section {
        clear:both;
        display:block;
        margin-left:20px;
    }

    #writing article ol,
    #writing article ul {
        margin-left:60px;
    }

    #writing article ol li,
    #writing article ul li {
        line-height:28px;
    }

    #writing article h3 {
        font-size:24px;
        line-height:28px;
        margin:28px 0;
    }

    p.tags {
        margin:-14px 0 42px;
    }

    p.tags strong {
        color:#8c8c8c;
    }
    

    /* ----- the footer ----- */

    footer {
        color:#8c8c8c;
        display:block;
        font-size:14px;
        line-height:28px;
        margin:0 0 80px 60px;
        width:600px;
    }

    footer nav {
        float:right;
    }

    /* ----- github ribbon ----- */

    .ribbon {
        border:0;
        position:absolute;
        right:0;
        top:0;
        z-index:100;
    }

    /* 
     *
     * Cleafloat: A Haiku
     * 
     * Markup zealots, please:
     * overflow:hidden is not
     * always an option.
     *
     */

    .clearfloat:after {
        display:block;
        visibility:hidden;
        clear:both;
        height:0;
        content:".";
    }
        
    .clearfloat {
        display:inline-block;
    }

    .clearfloat {
        display:block;
    }
    """,}
    return dict([(k, dedent(v).strip()) for k, v in templates.items()])

META_REGEX = re.compile(r"/(\d{4})\.(\d\d)\.(\d\d)\.(.+)")

if __name__ == "__main__":
    templates = get_templates()
    env = Environment(loader=DictLoader(templates))
    all_entries = read_and_parse_entries()
    shutil.copytree(DIRS['assets'], DIRS['build'])
    generate_index(all_entries, env.get_template('list.html'))
    os.mkdir(join(DIRS['build'], 'tags'))
    generate_tag_indices(all_entries, env.get_template('list.html'))
    generate_details(all_entries, env.get_template('detail.html'))
    generate_404(env.get_template('404.html'))
    generate_style(templates[STYLESHEET])
    shutil.rmtree(DIRS['public'])
    shutil.move(DIRS['build'], DIRS['public'])
